From fb68a06681615267c5932271c26990b1e3c36d4d Mon Sep 17 00:00:00 2001
Message-ID: <fb68a06681615267c5932271c26990b1e3c36d4d.1776946505.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Thu, 16 Apr 2026 19:30:46 +0200
Subject: [RFC PATCH 10/28] physmem: use RAMBlockMapOps for address_space
 caches

Teach address_space_cache_init() and address_space_cache_destroy() to
use RAMBlockMapOps for provider-backed RAMBlocks.

Also keep direct-map-only caches empty on non-direct translations and
reset cache state through address_space_cache_init_empty().

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 system/physmem.c | 27 ++++++++++++++++++++++-----
 1 file changed, 22 insertions(+), 5 deletions(-)

diff --git a/system/physmem.c b/system/physmem.c
index a849b78bed2b..026a505b74c9 100644
--- a/system/physmem.c
+++ b/system/physmem.c
@@ -3859,9 +3859,9 @@ void address_space_unmap(AddressSpace *as, void *buffer, hwaddr len,
         return;
     }
 
-
     BounceBuffer *bounce = container_of(buffer, BounceBuffer, buffer);
     assert(bounce->magic == BOUNCE_BUFFER_MAGIC);
+    (void)bounce;
 
     if (is_write) {
         address_space_write(as, bounce->addr, MEMTXATTRS_UNSPECIFIED,
@@ -3911,6 +3911,7 @@ int64_t address_space_cache_init(MemoryRegionCache *cache,
     Int128 diff;
 
     assert(len > 0);
+    address_space_cache_init_empty(cache);
 
     l = len;
     cache->fv = address_space_get_flatview(as);
@@ -3937,8 +3938,14 @@ int64_t address_space_cache_init(MemoryRegionCache *cache,
                                         cache->xlat, l, is_write,
                                         MEMTXATTRS_UNSPECIFIED);
         cache->ptr = qemu_ram_ptr_length(mr->ram_block, cache->xlat, &l, true,
-                                         is_write);
+                                         is_write, RAM_BLOCK_MAP_KIND_CACHE);
     } else {
+        if (as->direct_map_only) {
+            memory_region_unref(mr);
+            flatview_unref(cache->fv);
+            address_space_cache_init_empty(cache);
+            return 0;
+        }
         cache->ptr = NULL;
     }
 
@@ -3962,14 +3969,24 @@ void address_space_cache_destroy(MemoryRegionCache *cache)
     if (!cache->mrs.mr) {
         return;
     }
+    /*
+     * cache->mrs.mr was validated above, and init_empty() normalization keeps
+     * cache->ptr/cache->xlat/cache->len safe to dereference for teardown.
+     */
 
-    if (xen_enabled()) {
+    if (cache->ptr && cache->mrs.mr->ram_map_ops && cache->mrs.mr->ram_block &&
+        cache->mrs.mr->ram_block->host == NULL) {
+        cache->mrs.mr->ram_map_ops->put(cache->mrs.mr->ram_map_opaque,
+                                        cache->mrs.mr->ram_block,
+                                        cache->ptr, cache->xlat, cache->len,
+                                        cache->is_write,
+                                        RAM_BLOCK_MAP_KIND_CACHE);
+    } else if (xen_enabled()) {
         xen_invalidate_map_cache_entry(cache->ptr);
     }
     memory_region_unref(cache->mrs.mr);
     flatview_unref(cache->fv);
-    cache->mrs.mr = NULL;
-    cache->fv = NULL;
+    address_space_cache_init_empty(cache);
 }
 
 /* Called from RCU critical section.  This function has the same
-- 
2.53.0

